service/history/queues/slice.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	incomingSlice, ok := slice.(*SliceImpl)
+	if !ok {
+		panic(fmt.Sprintf("Unable to merge queue slice of type %T with type %T", s, slice))
+	}
+
+	s.destroy()
+	incomingSlice.destroy()
+	return &SliceImpl{
+		executableInitializer:  s.executableInitializer,
+		scope:                  s.scope.MergeByPredicate(incomingSlice.scope),
+		outstandingExecutables: s.mergeExecutables(incomingSlice),
+		iterators:              s.mergeIterators(incomingSlice),
+	}
+}
+
+func (s *SliceImpl) mergeExecutables(incomingSlice *SliceImpl) map[tasks.Key]Executable {
+	thisExecutables := s.outstandingExecutables
+	thatExecutables := incomingSlice.outstandingExecutables
+	if len(thisExecutables) < len(thatExecutables) {
+		thisExecutables, thatExecutables = thatExecutables, thisExecutables
+	}
+
+	for key, executable := range thatExecutables {
+		thisExecutables[key] = executable
+	}
+
+	return thisExecutables
+}
+
+func (s *SliceImpl) mergeIterators(incomingSlice *SliceImpl) []Iterator {
+	mergedIterators := make([]Iterator, 0, len(s.iterators)+len(incomingSlice.iterators))
+	currentIterIdx := 0
+	incomingIterIdx := 0
+	for currentIterIdx < len(s.iterators) && incomingIterIdx < len(incomingSlice.iterators) {
+		currentIter := s.iterators[currentIterIdx]
+		incomingIter := incomingSlice.iterators[incomingIterIdx]
+
+		if currentIter.Range().InclusiveMin.CompareTo(incomingIter.Range().InclusiveMin) < 0 {
+			mergedIterators = s.appendIterator(mergedIterators, currentIter)
+			currentIterIdx++
+		} else {
+			mergedIterators = s.appendIterator(mergedIterators, incomingIter)
+			incomingIterIdx++
+		}
+	}
+
+	for _, iterator := range s.iterators[currentIterIdx:] {
+		mergedIterators = s.appendIterator(mergedIterators, iterator)
+	}
+	for _, iterator := range incomingSlice.iterators[incomingIterIdx:] {
+		mergedIterators = s.appendIterator(mergedIterators, iterator)
+	}
+
+	return mergedIterators
+}
+
+func (s *SliceImpl) appendIterator(
+	iterators []Iterator,
+	iterator Iterator,
+) []Iterator {
+	if len(iterators) == 0 {
+		return []Iterator{iterator}
+	}
+
+	size := len(iterators)
+	if iterators[size-1].CanMerge(iterator) {
+		iterators[size-1] = iterators[size-1].Merge(iterator)
+		return iterators
+	}
+
+	return append(iterators, iterator)
+}
+
+func (s *SliceImpl) ShrinkRange() {
+	s.validateNotDestroyed()
+
+	minTaskKey := tasks.MaximumKey
+	for key := range s.outstandingExecutables {
+		if s.outstandingExecutables[key].State() == ctasks.TaskStateAcked {
+			delete(s.outstandingExecutables, key)
+			continue
+		}
+
+		minTaskKey = tasks.MinKey(minTaskKey, key)
+	}
+
+	// still has pending task in memory
+	if len(s.outstandingExecutables) != 0 {
+		s.scope.Range.InclusiveMin = minTaskKey
+		return
+	}
+
+	// no more pending task in memory, but has more tasks to read from persistence
+	if len(s.iterators) != 0 {
+		s.scope.Range.InclusiveMin = s.iterators[0].Range().InclusiveMin
+		return
+	}
```

Comment:
this logic depends on 
`in mem tasks keys <= all iterator ranges` 

is this invariant always true?

Reply:
Yes, iterator range is updated when returning task, so its range only contains tasks that's still in db. and since tasks are loaded in order, in mem task keys are always < iterator ranges

Reply:
what if 

slice 1: some in mem tasks & iter
slice 2: some in mem task & iter

merge slice 1 & 2?

Reply:
Ah good point. The issue is on L356 actually. 
When there's still pending tasks in memory, I need to pick the smaller between min pending task key and iterators[0].Range().Min (if exists).
When there's no pending tasks in memory, the logic on L363 still holds.

Reply:
Fixed. Updated test to validate slice state after shrink range as well.